hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/UgiFilter.java

```diff
@@ -0,0 +1,169 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.ozone.s3;
+
+import org.apache.hadoop.hdds.conf.OzoneConfiguration;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.ozone.OzoneSecurityUtil;
+import org.apache.hadoop.ozone.s3.exception.OS3Exception;
+import org.apache.hadoop.ozone.s3.signature.SignatureInfo;
+import org.apache.hadoop.ozone.s3.signature.StringToSignProducer;
+import org.apache.hadoop.ozone.security.OzoneTokenIdentifier;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.security.token.Token;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.WebApplicationException;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.security.PrivilegedExceptionAction;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Enumeration;
+import org.apache.hadoop.ozone.s3.signature.AWSSignatureProcessor;
+
+import com.google.common.annotations.VisibleForTesting;
+import static org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.OMTokenProto.Type.S3AUTHINFO;
+import static org.apache.hadoop.ozone.s3.exception.S3ErrorTable.MALFORMED_HEADER;
+
+/**
+ * Preprocessing filter for every request.
+ * - creates OzoneToken containing aws signature
+ * aws id and stringToSign for aws authenication;  stores OzoneToken in
+ * thread local variable (UserGroupInformation object) avail to all
+ * s3 rest command endpoints
+ */
+public class UgiFilter implements Filter {
+  public static final Logger LOG = LoggerFactory.getLogger(UgiFilter.class);
+
+  @Inject
+  private OzoneConfiguration ozoneConfiguration;
+  @Inject
+  private Text omService;
+
+  @Override
+  public void init(FilterConfig filterConfig) throws ServletException {
+
+  }
+
+  @Override
+  public void doFilter(ServletRequest servletRequest,
+                       ServletResponse servletResponse, FilterChain filterChain)
+      throws IOException, ServletException {
+    Map<String, String> headerMap = new HashMap<>();
```

Comment:
Question:
Previously we used to use context.getHeaders() which is multivalued map. So, for a single key it can have multiple values. Now with this logic of Map, if a key has multiple values.

And as we use getHeader which returns first one only. Might be here we need to use getHeaders() and use multi map or use   @Context private ContainerRequestContext context; as before
JavaDoc says below.
`If there are multiple headers with the same name, this method returns the first head in the request. `

Reply:
> Question:
> Previously we used to use context.getHeaders() which is multivalued map. So, for a single key it can have multiple values. Now with this logic of Map, if a key has multiple values.

Thanks @bharatviswa504. Right, previously we had a multivalued map from getHeaders() however we only used the first value for each header. (AWSSignatureProcessor fromHeaderMap() headerEntry.getValue().get(0) )

Here we use getHeader() as you wrote that is mapped in the same manner (first value entry). Also, for the injected ContainerRequestContext, we removed the injected context because it is unavailable when generating the stringToSign for the ozone S3AUTHINFO token in UgiFilter.

For aws authentication, (stringToSign), is there a need for us to extend the implementation to support multi-valued headers? Right now the PR implementation should provide the same as the prior version (AWSSignatureProcessor) using the single header value.

Reply:
Yes you are right. I think we should be fine with this.